doc: extensions: boards: add ability to filter by hw capability

kartben opened the pull request:
Augment the board catalog with the list of supported hardware features for each board, by building -- cmake only, as we only need to get to the point where we obtain a `zephyr.dts` -- hello world for all boards and the analyzing the contents of the resulting Devicetree.

The board index page search form has been updated to allow to filter by supported HW capability, e.g. what board supported in zephyr has "display + lora + i2s"

A follow-up PR will leverage the metadata to automatically populate the currently very manual, often stale, table of "Supported Features" that boards' READMEs typically include (folks might remember seeing a demo of this -- I decided to push it to a different PR to simplify review process)

<img width="864" alt="image" src="https://github.com/user-attachments/assets/6fa0ac63-5a3e-4e75-9954-5512be84bfd3" />

Doc GH workflow running on pull requests will skip the build of the 800+ hello worlds as it is quite resource-heavy and only the scheduled build (running every 3 hours) will do it (similar to how we skip generation of the fulls docs of the devicetree bindings).

Related: due to this PR updating the doc-build.yml workflow, it's not publishing the CI output to builds.zephyrproject.io ; best way to test the actual user experience is to get the most recent successful build from my fork ([kartben/zephyr/actions/workflows/doc-build.yml?query=event%3Aschedule](https://github.com/kartben/zephyr/actions/workflows/doc-build.yml?query=event%3Aschedule)), download the html-output artifact, unzip it, and open index.html in a browser

kartben committed:
doc: extensions: boards: Add hardware features filter to board catalog

Add a new hardware features filter to the board catalog that allows
users to filter boards based on their supported hardware capabilities.
The features are extracted from the devicetree files and organized by
binding type.

Key changes:
- Extract hardware feature descriptions from devicetree bindings
- Add HW_FEATURES_TURBO_MODE option to skip feature generation for
  faster builds (similar to DT_TURBO_MODE)
- Add tag-based UI for filtering boards by hardware features

The feature can be disabled for faster documentation builds using
-DHW_FEATURES_TURBO_MODE=1 or by using 'make html-fast'.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

pdgendt commented on doc/_extensions/zephyr/domain/__init__.py:
```diff
@@ -70,6 +69,145 @@
+    "watchdog": nodes.Text("Watchdog"),
+    "wifi": nodes.Text("Wi-Fi"),
+    "xen": nodes.Text("Xen"),
+    "xspi": nodes.abbreviation("XSPI", "XSPI", explanation="Expanded Serial Peripheral Interface"),
```
This large sorted list with some multiline blocks was the inspiration for #80444

Only check lines that match a regex, for example:
```python
    # zephyr-keep-sorted-start /^    "/
    # ...
    # zephyr-keep-sorted-stop
```

kartben replied:
Ahah yes, *if* this stays I will definitely flag it with the tags, thanks!
Right now I am actually very much interested in --although of course PR is still in draft, ... but since you're here 😄-- figuring out what makes for a "good" categorization of the supported HW feature. As you may have seen from the current implementation, this is making every directory under dts/bindings/ a category, but having ~120 categories doesn't seem right and there's probably a better way? Maybe we keep the approach but do some addition grouping? Like, dunno, off the top of my head, grouping `fuel-gauge` and `battery` ("Battery mgmt"), or `usb` and `usb-c` (USB), ... together ?

pdgendt replied:
Oopf, 120 is a lot, I think grouping would make sense, also to have some better names than for example `mm`.

kartben replied:
will be moved to a different PR - thanks!

erwango commented:
@kartben Would it be possible to add a way to see what features are supported for a given SoC series ?

kartben commented:
> @kartben Would it be possible to add a way to see what features are supported for a given SoC series ?

@erwango I *think* we need some kind of "board" for this to work. But if there is a way to feed a SoC DTS into [this](https://docs.zephyrproject.org/latest/build/dts/intro-input-output.html) then we'd be in business

kartben committed:
samples: hello_world: set min RAM to 2KB for Hello World test

It is possible to fit the Hello World in 2K of RAM so update the
sample.yaml file accordingly. It might not *run*, but then the console
harness will catch that.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
ci: doc: add action-zephyr-setup

Setup a fully working Zephyr workspace so that documentation steps that
require e.g. building Zephyr apps have the ability to do so.
The "turbo" builds running on pull requests get a minimal setup, with
no toolchain being installed.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
twister: allowlist build_info.yml and zephyr.dts

These files can be useful to troubleshoot a test that's failing,
and they're really small, too.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben commented on boards/index.rst:
```diff
@@ -24,6 +24,10 @@ this page <boards-shields>`.
      single field, selecting multiple options (such as two architectures) will show boards matching
      **either** option.
 
+   * The list of supported hardware features for each board is automatically generated using
```
is it not?

<img width="564" alt="image" src="https://github.com/user-attachments/assets/f7cfaf42-5220-4a12-9cb0-7335e3f37a3f" />

https://github.com/zephyrproject-rtos/zephyr/pull/79754/files#diff-94a95bd3ad716e01fb3a81b44a4be5b9eed1fc1342b1fdfb86200ef4100cb84eR128

tobiaskaestner commented:
One thing that might be worth considering is that you now also need a full install of the SDK now to build the docs without error. What I mean by this is that I get a lot of build errors one the helloworld sample that is used to create the devicetree simply because I do not have all toolchains installed (in my case I typically install arm and x86_64 and sometimes riscV but none of the others). Perhaps it should be mentioned somewhere.
Alternatively I am thinking that if all we need is the compiled devicetree we could craft an artificial sample for this purpose that leverages the find_package(Zephyr COMPONENTS dts) mechanism to only run the the cmake build system up to this point (which is before CMake is trying to figure out the cross-compiler to use).

kartben commented:
Pushed a [minor fix](https://github.com/zephyrproject-rtos/zephyr/compare/89ac55271bb4be66bd34f12b22918ab9421c872a..632aeb3c48d5b57936eca0d14e4e789f538404ed)
The `action-zephyr-setup` doesn't really enjoy not installing any toolchain (more specifically, it confused github actions/cache). Restoring the full blown "all" cache is actually really fast, so let's keep things simple.

kartben commented:
> One thing that might be worth considering is that you now also need a full install of the SDK to build the docs without error. What I mean by this is that I get a lot of build errors on the helloworld sample that is used to create the devicetree simply because I do not have all toolchains installed (in my case I typically install arm and x86_64 and sometimes riscV but none of the others). Perhaps it should be mentioned somewhere. Alternatively I am thinking that if all we need is the compiled devicetree we could craft an artificial sample for this purpose that leverages the find_package(Zephyr COMPONENTS dts) mechanism to only run the the cmake build system up to this point (which is before CMake is trying to figure out the cross-compiler to use).

@tobiaskaestner 
Ya. I tried to hint at it in https://github.com/zephyrproject-rtos/zephyr/pull/79754/files#diff-96b89ca373105b69b0003ffc40ce2715c6fd7dd8a0a3c25e44aeaf1227f3ae8f

> Alternatively I am thinking that if all we need is the compiled devicetree we could craft an artificial sample for this purpose that leverages the find_package(Zephyr COMPONENTS dts) mechanism to only run the the cmake build system up to this point (which is before CMake is trying to figure out the cross-compiler to use).

That's a great idea. I don't know enough about the build system to know if this is a possibility or how easy that would be. Would you be able to give it a try?

golowanow reviewed (dismissed):
Cool feature, although it takes huge time to build docs now: ~3x longer than html-fast. Perhaps it is worth to mention at https://docs.zephyrproject.org/latest/contribute/documentation/generation.html#running-the-documentation-processors

kartben commented:
> Alternatively I am thinking that if all we need is the compiled devicetree we could craft an artificial sample for this purpose that leverages the find_package(Zephyr COMPONENTS dts) mechanism to only run the the cmake build system up to this point (which is before CMake is trying to figure out the cross-compiler to use).

something like this... possibly?

```
# SPDX-License-Identifier: Apache-2.0

cmake_minimum_required(VERSION 3.20.0)

find_package(Zephyr COMPONENTS FindZephyr-sdk zephyr_default yaml pre_dt dts REQUIRED HINTS $ENV{ZEPHYR_BASE})

project(hello_world)

target_sources(app PRIVATE src/main.c)
```

Need to do more testing but it's not clear it's really faster though :|

mbolivar commented on doc/_scripts/gen_boards_catalog.py:
```diff
@@ -51,7 +56,98 @@ def guess_doc_page(board_or_shield):
+                    board_target = f"{board_target}@{revision}"
+
+                # Create EDT object for this board's devicetree
+                try:
+                    edt = edtlib.EDT(
+                        str(dts_file),
+                        [str(ZEPHYR_BASE / "dts/bindings")],
+                        warn_reg_unit_address_mismatch=False
+                    )
```
It would be better to load edt.pickle instead, that's what it's there for. The default build environment can in the general case get a different set of EDT constructor arguments than this, so you can't trust that you're faithfully recreating the original. See scripts/west_commands/sign.py or scripts/build/check_init_priorities.py among several other places for examples of how to load this object.

(Note https://docs.zephyrproject.org/latest/build/dts/intro-input-output.html mentions zephyr.dts is purely for debugging for this reason)

kartben replied:
Thanks for the review! The pickle is not part of the artifacts still left behind by Twister when running in "-M" mode, which I am to keep disk usage low enough to fit in the limits of a default GitHub-hosted runner. Will see what can be done but I am wondering if there are any differences we can reasonably expect to see on a "vanilla" hello world build (but I do get your point, don't get me wrong, and using the pickle would probably be better—and likely faster). Like, for example, it's not like there would be additional binding_dirs or vendor_prefixes to be aware of?

kartben replied:
FWIW I can confirm I get exact same output with the pickle and when loading the EDT from zephyr.dts.
I guess I could get the list of binding_dirs from the `build_info.yml` file rather than hardcoding it, though.

mbolivar replied:
If you insist, I'll drop the -1, but I'm really uncomfortable with this. The edt.pickle file went in exactly for this purpose and it exists as an abstraction barrier to prevent scripts like this from breaking as we make changes to edtlib.

kartben replied:
> If you insist, I'll drop the -1, but I'm really uncomfortable with this. The edt.pickle file went in exactly for this purpose and it exists as an abstraction barrier to prevent scripts like this from breaking as we make changes to edtlib.

@mbolivar
I'll take you up on your offer and will insist then, if that's ok :)
As you can imagine this is only the first iteration and I'll be happy to look at using the edt.pickle file to improve robustness/accuracy

henrikbrixandersen replied:
I must admit I think the fact that the devicetree maintainer is uncomfortable with the use of `zephyr.dts` over `edt.pickle` is enough to keep this from going in at this point in the v4.1 release cycle.

mbolivar reviewed (dismissed):
Looks great! I'd like to make sure data this feature sees is exactly the same as what gen_defines.py created, so please can you consume the edt.pickle object that exists for this reason instead of trying to reparse the DTS?

kartben committed:
twister: Add --keep-artifacts option

Introduce a new command-line option `--keep-artifacts` in twister that
allows users to specify which artifacts should be preserved
during test cleanup in addition to the default set.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>
Co-authored-by: Anas Nashif <anas.nashif@intel.com>

kartben commented:
Pushed to the wrong branch 😢  sorry @nashif and @erwango I dismissed your reviews. The branch is now back to the exact same commit you'd approved ([28c696b](https://github.com/zephyrproject-rtos/zephyr/commit/28c696bb99c7f99e9532eb44cde7ddbf4845c62c))
